### ISSUE TITLE
Spectral Resonance Optimization for CRISPR gRNA Prediction

### ISSUE THREAD

**Issue by zfifteen** [zfifteen/unified-framework#905]
User Story: Spectral Resonance Optimization for CRISPR gRNA Prediction (φ-phase + arcsin bridge)

Title

Optimize CRISPR gRNA prediction by applying a golden-ratio phase transform \theta’(n,k) with arcsin phase compression to complex DNA waveforms, validating gains over RuleSet3 baselines and enabling invariant-weighted ML pipelines.

Background
	•	Working repo implementing FFT-based DNA signal analysis with complex encoding (A=1, T=−1, C=+i, G=−i), φ-phase \theta’(n,k)=\varphi\big((n\bmod \varphi)/\varphi\big)^k, arcsin bridging, and benchmarking scripts: zfifteen/wave-crispr-signal.  ￼
	•	Rule Set 3 (RS3) is the standard on-target baseline for sgRNA activity (Nature Communications 2022; DeWeirdt et al.) with a maintained reference implementation.  ￼

Problem Statement

Current on-target predictors (e.g., RS3) do not exploit spectral structure induced by complex waveform encodings and position-dependent φ-phase morphologies. We hypothesize that a φ-phase transform tuned near k^*\!\approx\!0.300 with arcsin compression sharpens fundamental spectral energy, suppresses sidelobes, and improves composite disruption scoring, yielding measurable AUC lift vs RS3 on public CRISPR datasets.

Goals
	1.	Implement φ-phase and arcsin bridge transforms over complex DNA waveforms; extract spectral features (ΔEntropy, Δf₁, sidelobe ratios) and composite disruption scores.
	2.	Benchmark against RS3 using the repo’s validation harness with fixed seeds and bootstrap CIs; report ΔAUC and GC-quartile correlations.
	3.	Integrate spectral features as invariant weights into a model-agnostic ML pipeline for variant prioritization; document ablations and failure modes.
	4.	Produce reproducible artifacts (scripts, JSON results, CI checks) and a concise METHODS document suitable for manuscript integration.

Scope & Non-Goals
	•	In-scope: On-target efficiency prediction; spectral-distance metrics for off-target profiling; entropy-gradient features for repair-pathway forecasting.
	•	Out-of-scope: Editing outcome simulation beyond entropy-gradient correlates; wet-lab protocol design.

User Value
	•	Increased predictive accuracy over RS3 with interpretable spectral features.
	•	Portable, invariant-weighted features usable with multiple model families (logistic/XGBoost/NN).
	•	Reproducible pipeline for other researchers to verify, extend, and compare.

Requirements

Functional
	•	Complex encoding: Map A/T/C/G → 1,-1,+i,-i; optional context padding and windowing (Hann).
	•	φ-phase transform: \theta’(n,k)=\varphi((n\bmod \varphi)/\varphi)^k, applied as y[n;k]=x[n]\exp(i\theta’); sweep k\in[0.20,0.40] at Δk=0.005; default k^*=0.300.
	•	Arcsin bridge: \tilde{z}=\arcsin(\alpha\sin z) with z=\operatorname{Arg}(x)+\theta’, 0.85\le\alpha\le0.98; default \alpha=0.95.
	•	FFT & features: Compute Y=\mathrm{FFT}(y); derive ΔEntropy, Δf₁ (peak/centroid near f_1\approx N/L), and sidelobe ratios; assemble a composite disruption score D via calibrated weights.
	•	Off-target spectral distance: Magnitude-squared coherence proxy and Wasserstein-1 on normalized spectra; sidelobe asymmetry score.
	•	GC-quartiles: Compute per-guide GC%, assign quartiles, and correlate with D (Pearson/Spearman; FDR).
	•	Benchmarks vs RS3: Use RS3 scores as baseline (paper & package links below); compute ΔAUC with 10k stratified bootstrap; log seeds and environment.  ￼

Non-Functional
	•	Reproducibility: Fixed seeds; deterministic FFT; captured config in versioned YAML; CI job that reruns benchmarks and fails if ΔAUC < target threshold.
	•	Transparency: Persist per-guide features and predictions (CSV/Parquet) and a machine-readable results.json with metrics and CIs.

Deliverables
	1.	Code additions (to wave-crispr-signal):
	•	modules/phi_phase.py: φ-phase generator and arcsin bridge ops.
	•	modules/spectral_features.py: ΔEntropy/Δf₁/sidelobe/MSC/W₁/SAS.
	•	applications/cli.py: CLI for transform + feature extraction.
	•	proof_pack/run_validation.py: end-to-end RS3 comparison (already present—extend flags).  ￼
	2.	METHODS.md (derivations, hyperparameters, windows, band definitions, stats tests).
	3.	results/ artifacts:
	•	k_sweep_summary.csv, gc_quartile_corr.csv, auc_bootstrap.json (mean, 95% CI).
	4.	CI: GitHub Actions workflow to execute validation with pinned seeds and publish artifacts.
	5.	Ablations: No-arcsin, no-φ-phase, alternate A/T/C/G encodings, window variants, no invariant-weighting.

Acceptance Criteria
	•	AC-1: Running proof_pack/run_validation.py with defaults (k*=0.300, α=0.95, Hann, N∈{256,512}) produces a metrics bundle in results/ including ΔAUC vs RS3, ΔEntropy, Δf₁, sidelobe deltas, and GC-quartile correlations, with seeds logged.  ￼
	•	AC-2: ΔAUC vs RS3 is computed with 10k stratified bootstrap and reported with a 95% CI; script exits non-zero if artifacts are missing or seeds aren’t recorded.
	•	AC-3: GC-quartile correlation statistics (Pearson/Spearman with FDR) are emitted and saved to gc_quartile_corr.csv.
	•	AC-4: Off-target spectral-distance metrics (MSC and W₁) are available as features/penalties and serialized per guide.
	•	AC-5: METHODS.md fully specifies equations, windows, bands, and statistics; running pytest -q passes on a clean checkout.  ￼

Implementation Plan (links)
	•	Primary repo & scripts:
	•	zfifteen/wave-crispr-signal — codebase, docs, and proof pack (proof_pack/run_validation.py).  ￼
	•	Baseline model (Rule Set 3):
	•	Paper — Nature Communications (2022): “Accounting for small variations in the tracrRNA sequence improves sgRNA activity predictions for CRISPR screening.”  ￼
	•	Package — gpp-rnd/rs3 (reference implementation).  ￼
	•	Historical context — Rule Set 2 (Doench et al., 2016) for background comparisons.  ￼
	•	Context & constraints from related work:
	•	Project challenge & factoring context captured in zfifteen/z-sandbox#221 (method rigor ethos for reproducibility discipline across repos).  ￼

Configuration (example, checked-in as configs/k300.yaml)

encoding: {A: "1+0j", T: "-1+0j", C: "0+1j", G: "0-1j"}
window: hann
N: 256
phi_phase:
  enabled: true
  k: 0.300
arcsin_bridge:
  enabled: true
  alpha: 0.95
fft:
  pad_to_pow2: true
metrics:
  include: [delta_entropy, delta_f1, sidelobe_ratio, msc, wasserstein, sas]
bootstrap:
  n_draws: 10000
  stratify_by: locus
random_seeds:
  python: 1337
  numpy: 1337
  torch: 1337
baseline:
  ruleset3:
    provider: rs3
    version: pinned
outputs_dir: results/

CLI (example)

# Baseline vs φ-phase + arcsin (reproducible run)
python proof_pack/run_validation.py \
  --dataset data/guides.tsv \
  --k0 0.000 \
  --kstar 0.300 \
  --alpha 0.95 \
  --window hann \
  --N 256 \
  --bootstrap 10000 \
  --metrics auc,delta_entropy,delta_f1,sidelobe,coherence,wasserstein \
  --compare rs3 \
  --out results/k300_alpha095_seed1337.json

Data & Statistics
	•	Datasets: Use the repository’s documented public sets (see repo README “Validation & Proof Pack”) and ensure licensing compliance.  ￼
	•	Statistics: Stratified bootstrap for ROC-AUC; Pearson/Spearman with FDR for GC-quartiles; store all random seeds and environment (Python/NumPy/PyTorch versions) in artifacts.

Risks & Mitigations
	•	Risk: Over-compression (\alpha\to1) can flatten discriminative harmonics.
Mitigation: Monitor sign of Δsidelobe; add early-stop if Δsidelobe flips positive.
	•	Risk: GC-extremes confound entropy metrics.
Mitigation: Include GC% as covariate; report partial correlations.
	•	Risk: Window/length sensitivity.
Mitigation: Validate across {128,256,512} and report stability ranges.

Definition of Done
	•	CI run produces results/ with ΔAUC (bootstrap 10k), GC-quartile correlation stats (FDR), and per-guide feature tables.
	•	METHODS.md and configs are sufficient for third-party reproduction on a clean machine.
	•	Ablation report included in results/ablations/.

⸻

Reasoning (for Reproducibility by Other LLMs)
	1.	Use only repository-grounded claims and standard baselines: All implementation specifics (complex encoding, φ-phase, arcsin bridge, FFT features, proof-pack scripts) are anchored to zfifteen/wave-crispr-signal. Avoid inventing datasets or unpublished metrics; when stating results (e.g., ΔAUC, GC-quartile r/p), attribute to the repo’s validation outputs and ensure the benchmark scripts emit machine-readable artifacts.  ￼
	2.	Baseline provenance: Reference Rule Set 3 via its peer-reviewed paper and official package to ensure comparability and verifiable setup (model version pinning, tracrRNA considerations).  ￼
	3.	Determinism: Fix seeds (Python/NumPy/Torch), pin library versions, and record FFT window/length, zero-padding, and band definitions in configs and results.
	4.	Statistical rigor: Use stratified bootstrap (10k) for AUC; report 95% CI; for multiple feature tests (e.g., GC-quartiles), apply FDR and serialize full p-value tables.
	5.	Ablations to localize gains: Compare (a) φ-phase only, (b) arcsin only, (c) alternate encodings/windows, (d) removal of invariant weighting. Gains should persist primarily from φ-phase + arcsin interaction.
	6.	Artifacts first: Treat the CI-produced results/*.json and *.csv as the single source of truth for any reported numbers; any narrative must be derivable from these artifacts.

⸻

References & Links
	•	Primary repo: https://github.com/zfifteen/wave-crispr-signal (code, docs, proof pack).  ￼
	•	Rule Set 3 (paper): Nature Communications 2022 — DeWeirdt et al., “Accounting for small variations in the tracrRNA sequence improves sgRNA activity predictions for CRISPR screening.”  ￼
	•	Rule Set 3 (package): https://github.com/gpp-rnd/rs3.  ￼
	•	Rule Set 2 background (open-access): Doench et al., Nature Biotechnology 2016 (Azimuth/RS2).  ￼
	•	Context on rigor (team ethos): Related engineering discipline and challenge framing documented in https://github.com/zfifteen/z-sandbox/issues/221.  ￼
